accord-core/src/test/java/accord/burn/random/FrequentLargeRange.java

```diff
@@ -18,59 +18,121 @@
 
 package accord.burn.random;
 
-import accord.utils.Invariants;
+import accord.utils.Gen;
+import accord.utils.Gen.LongGen;
+import accord.utils.Gens;
 import accord.utils.RandomSource;
 
-public class FrequentLargeRange implements RandomLong
+import java.time.Duration;
+import java.util.concurrent.TimeUnit;
+
+public class FrequentLargeRange implements LongGen
 {
-    private final RandomLong small, large;
-    private final double ratio;
-    private final int steps;
-    private final double lower, upper;
-    private int run = -1;
-    private long smallCount = 0, largeCount = 0;
-
-    public FrequentLargeRange(RandomLong small, RandomLong large, double ratio)
+    private final LongGen small, large;
+    private final Gen<Boolean> runs;
+
+    public FrequentLargeRange(LongGen small, LongGen large, double ratio)
     {
-        Invariants.checkArgument(ratio > 0 && ratio <= 1);
         this.small = small;
         this.large = large;
-        this.ratio = ratio;
-        this.steps = (int) (1 / ratio);
-        this.lower = ratio * .8;
-        this.upper = ratio * 1.2;
+        this.runs = Gens.bools().biasedRepeatingRuns(ratio);
```

Comment:
Didn't we discus moving this to `RandomSource`?

Reply:
no, you want that and I strongly disagree with that.  If we keep pushing more and more generators into the `RandomSource` it gets too large and harder to discover things.  It also leads to a coupling I strongly feel is bad that everything is "owned" by a `RandomSource`, making it harder for reuse

Reply:
> Is there a reason some (e.g. pick) methods warrant entry to RandomSource but not others?

some picks have state, some do not; some are light weight, some are not.  I have been moving the picks with state outside as they need some extra state and a type to hold it.  so `pick(list)` can stay in the source as its stateless, but `pickWithWeight(list)` "should" stay outside as it needs to hold some state (the weight).  I actually don't mind moving `pick(list)` out into a static function, I just saw it as a very common pattern so felt it worked best in the interface in the current usage.

```
rs.pick(a, b, c)
// vs
pick(rs, a, b, c)
```

I am personally fine with both, but feel that the below is bad as it imposed a cost at each invoke, so makes more sense to push outside as it "should" return a holder of this state (Supplier, Gen, w/e).

```
rs.pickWithWeight(a, b, c)
```

feel free to call me out for my inconsistent stance with `pick(Set)`, this one violates what I stated above, and prob shouldn't be there; I mostly put it there out of laziness and to fix bugs I found with picking form sets in a non-deterministic manner... atm the function is dead code as its only used in my follow up work, so I am also cool with dropping it for now till its used.

> is why I created the abstraction

I am the author, I think you mean why you did similar in Simulator?  When I created this interface I wanted to make sure to have what you needed in simulator, what I needed in property tests, and perfectly match the JDK interface to make migration easier.  When I created this interface, I wanted to make sure it worked in all worlds and didn't impose anything on anyone...

> I think we're going to continue to have friction here. I will continue to embellish RandomSource as this is the natural place for me and how I create tests, and is why I created the abstraction. You will continue to put generators into Gen because this is how you prefer to write tests, and why you created that abstraction. This is fine while the approaches do not mix; we already have plenty of redundancy in test authorship. But we either have to accept that we have different approaches and stick to the approach already in use when editing tests, or we have to standardise.

This feels like a better conversation for dev@ or slack with a wider group of opinions; you prefer `Supplier`, I prefer `Gen`, but both work from a technical point of view so we need a tie breaking to move forward IMO.

>  I have generally tried to avoid imposing my views on your approach, as I don't see any reason to debate abstractions I don't intend to use.

We can choose to live in different worlds, but to be frank I am not the one imposing my views.  In the eviction branch I pointed out your property tests had issues with them and showed how to fix them but also mentioned you didn't need to write as the code existed; I let you choose which way to go.  In this patch your published stance is that I must conform to your view else I am blocked... 

> But in this case we're editing my work done with my approach

how are we editing *your* work when I am the author of this class?  This argument holds in https://github.com/apache/cassandra-accord/pull/58 as you are refactoring my work; this patch does not impose a `Gen` on *your* work; your patch does impose your views on me though.

> We seem to have as a result slipped into assuming we need to come to a conclusion about the one true approach, but I fear on this we will ever reach agreement. If we can agree not to impose our approaches on the other, and when editing tests to try to obey the existing style (as we do elsewhere in the codebase), I do not see why both (and more!) approaches, and generators, cannot co-exist.

I would prefer this, we even had a similar conversation in dev@ when SAI wanted to add a new random testing library to the project; the stance we moved forward with is allow all to live.

But to restate what I said above, I am not forcing my stuff on you, you are forcing yours on me.

Reply:
`RandomSource` is a port of my work for the simulator, and copied over at my suggestion.

`RandomDelayQueue`, `BurnTest` and `Cluster` are all my creations and these are where you introduced your new abstractions.

Reply:
> where you introduced your new abstractions.

* `RandomDelayQueue`/ `BurnTest` uses `accord.burn.random.FrequentLargeRange` which has existed for a long time at this point, and to follow the style you are imposing on me I convert to a `Supplier`, so in this patch the stored interface is a `Supplier` (again, the interface you are imposing), but all we do is reusing existing work.
* `Cluster` doesn't use `Gen` in this patch, all the random selection is delegated to the objects, which isn't a change to how this class works as this class never used `RandomSource` or anything directly, it delegates all randomness to its other elements; which again is unchanged in this patch.

Reply:
Cluster has been refactored to NodeSink, which alongside BurnTest and RandomDelayQueue were modified in this way prior to your reverting them, which happened after this debate was initiated.

Anyway, there’s no point debating this further if we’re in agreement about how to proceed for now. Though I fear we’re going to continue to revisit this.

Reply:
> Cluster has been refactored to NodeSink,

do you mean the moving the `partition` logic to `NodeSink`?  That doesn't have anything to do with `Gen` vs `Supplier` argument and was done to keep the logic clear and easier for debug logging; I added logging that shows if we chose to drop the msg vs we dropped due to it being partitioned.

> which alongside BurnTest and RandomDelayQueue were modified in this way prior to your reverting them, which happened after this debate was initiated.

* `BurnTest` always returned a Supplier, it was just a lambda now its a function call

```
return () -> time.next(rs);
// vs now
return time.asSupplier(rs);
```

The change in this patch was to reusing existing logic to implement a behavior we wanted, to add jitter for time... 

* `RandomDelayQueue` never exposed `Gen`, it just used an existing type `accord.burn.random.FrequentLargeRange` and called it using `.next`.  A better argument in this case is that the code was cleaner to extract the mapping logic to the constructor and use a `LongSupplier` in this case, so in this case I would agree the `LongSupplier` is cleaner code than the previous, hence why I didn't bother to push back here.

```
// before
TimeUnit.NANOSECONDS.toMillis(jitterNanos.next(rs))
// after
jitterMillis.getAsLong()
```

> Though I fear we’re going to continue to revisit this.

I agree I feel this is going to keep coming up.  The examples you gave never exposed `Gen`, they just reusing existing code... so the pushback to me was that when using existing code that happens to have a `Gen` interface, and intersects classes that you feel must be a given way, then must use a style you want, that is not documented, fleshed out, or voted on...  so as a contributor to this project I have no idea when the next argument will pop up.

What's also unexpected to me about this is that Simulator has a similar style to `Gen` (it has both styles); you supply the `RandomSource`[1].  We even spoke months back about one day unifying these so that Simulator extends the logic in Accord.  So from my view this feedback came out of no where and became a blocking comment... so when will the next one come up?

[1]

```
// see org.apache.cassandra.simulator.ClusterSimulation.Builder#schedulerFactory
static SchedulerFactory schedulerFactory(RunnableActionScheduler.Kind... kinds)
        {
            return (random) -> {
                switch (Choices.random(random, kinds).choose(random))
                {
                    default: throw new AssertionError();
                    case SEQUENTIAL: return new RunnableActionScheduler.Sequential();
                    case UNIFORM: return new RunnableActionScheduler.RandomUniform(random);
                    case RANDOM_WALK: return new RunnableActionScheduler.RandomWalk(random);
                }
            };
        }
```

Reply:
Perhaps we are talking past each other? Now we no longer utilise the `Gen` API within the top-level declarations of these classes, I am suggesting we move forwards. Your examples are comparing this patch in its current form with the patch in its form when this disagreement arose, which is to say where `Gen.LongGen` etc was peppered around the implementation of the aforementioned classes.

https://github.com/apache/cassandra-accord/pull/57/commits/49da37997ab4b1406c0069c9f387ab025d8990bb

This is mixing styles in my book. In fact, there remain examples I would prefer we removed, e.g. https://github.com/apache/cassandra-accord/pull/57/commits/49da37997ab4b1406c0069c9f387ab025d8990bb#diff-d7eab055034d6123ac548c3da83300a6ab4aa050e057d3d2cf03b059ab6df502R197 - this is very different to my style of authorship, which would directly compose primitive actions on `RandomSource` since that is entirely sufficient here. I haven't asked for that, though; just to isolate the usages to as close to the declaration as possible, and not confuse the API level.

My style of code does not use `Gen` at all, and introducing it - as demonstrated above - leads to a very different style of code being written. This is totally fine when authoring your own classes, but it's less well received by authors of their own classes that they're still maintaining, in which consistency of style will be harmed. I don't think this is the first time we've discussed our conflicting approaches to test authorship, so this shouldn't be a huge surprise? If your style and its components begin to weave their way amongst my style of authorship, then suddenly I have to have opinions about how your API is structured, it's inescapable.

Reply:
> Your examples are comparing this patch in its current form with the patch in its form when this disagreement arose, 

No, I am not.  I am replying to your comments and arguing they are not really valid; I give such examples as `return () -> time.next(rs);` vs `return time.asSupplier(rs);` which actively compares pre-feedback and post-feedback.  I also reply to your comment about `Cluster` and `NodeSink` where the change doesn't have anything to do with `Gen` vs `Supplier`, the change was mostly for debug logging reasons.

>  I am suggesting we move forwards.

What are the open issues before a +1 is given?  It is not clear what feedback is required vs optional as the only active threads I see is https://github.com/apache/cassandra-accord/pull/57#discussion_r1301845319 (small change to a single method) and this thread, so I want to make sure I understand the scope of work remaining.

> This is mixing styles in my book.

This is not at all true though.  `trunk` branch is currently 8c7a3c9ef4209d635b186189e17a2d9e728e9871, which has the following call paths that this PR touches that are under conflict

* `BurnTest` `nowSupplier`:  https://github.com/apache/cassandra-accord/blob/8c7a3c9ef4209d635b186189e17a2d9e728e9871/accord-core/src/test/java/accord/burn/BurnTest.java#L202-L206.  In this example we use a `Supplier<LongSupplier>` as we create one per node, and `Node` uses `LongSupplier` for time.  Within this supplier we do `return () -> Math.max(0, delayQueue.nowInMillis() + (jitter.nextInt(10) - 5));`.  Before applying the feedback you asked for, we still returned a `Supplier<LongSupplier>` and replaced the body to be `return () -> Math.max(0, queue.nowInMillis() + jitter.nextLong(forked));`. So, the only change before was

```
// before
jitter.nextInt(10) - 5
// after
jitter.nextLong(forked)
```

Before the logic was stateless, but we moved to a stateful model to express "runs" of long jitter, so moved from calling `RandomSource` to `FrequentLargeRange`.  I don't see how this logic was not within the style before and how it is now, or how future changes are any different.

* `RandomDelayQueue` `add`: https://github.com/apache/cassandra-accord/blob/8c7a3c9ef4209d635b186189e17a2d9e728e9871/accord-core/src/test/java/accord/impl/basic/RandomDelayQueue.java#L102

```
// before
random.nextInt(500)
// after
TimeUnit.NANOSECONDS.toMillis(jitterInNano.nextLong(random))
// after cleanup
jitterInMillis.nextLong(random)
```

Again, this moved from a stateless source of random to a stateful one and the local style is not really changed.

* `NodeSink` `send`: https://github.com/apache/cassandra-accord/blob/8c7a3c9ef4209d635b186189e17a2d9e728e9871/accord-core/src/test/java/accord/impl/basic/NodeSink.java#L65-L79. This class would delegate delivery and timeouts to the `PendingQueue` and only used random to change the delay; which is mostly unchanged in my patch (before/after).  The core change to this class is making it so we don't `DELIVER` every time and conditionally deliver; which is the feature request that this JIRA is scoped to do.  So

```
// before 
parent.add(self, to, messageId, send); // deliver
// after
Action action = partitioned(to) ? Action.DROP_PARTITIONED
                                        // call actions() per node so each one has different "runs" state
                                        : nodeActions.computeIfAbsent(to, ignore -> actions()).next(rs);
// can be a DELIVER, DELIVER_WITH_FAILURE, FAILURE, DROP, or DROP_PARTITIONED
```

The other change was that we don't use "cpu scheduling delay" (as we talked about) and to use a "network latency delay", so again

```
// before
parent.add(self, to, messageId, send); // deliver
// after
parent.add(packet, networkJitterNanos(to), TimeUnit.NANOSECONDS);
```

I struggle to see how I change the "style" (outside of `NodeSink` which went from nothing to something...).  If this is the argument imposed in this PR I have no clue how to deal with https://github.com/apache/cassandra-accord/pull/58 which actually changes the style (moving away from `RandomSource` in favor of `Supplier` and interfaces that extend it.  This moves away from the top level classes *owning* the random and how to do things in favor of a new abstraction)!

> this is very different to my style of authorship

I have to push back on this... Apache Accord style guide is the Apache Cassandra style guide, which does not document what you are imposing; that is your personal style.  Our current rules state that you have to have `clearly expressed and reasonable technical grounds` for such a pushback; we are supposed to work as a team.  If we randomly start changing the rules without voting and documenting, then we are not able to work as a team.

> since that is entirely sufficient here

Your patch exposes a new interface and move away from `RandomSource` within the context of this argument (since we are talking about the code calling `.next`)... you use `RandomSource` to create a `Supplier`, which is a new style you wish to move forward with in https://github.com/apache/cassandra-accord/pull/58.  In this patch I use `FrequentLargeRange`, so is your argument here that `FrequentLargeRange` should be merged into a function call within `RandomSource`?  If so this is the first time I have seen such an argument imposed here, the closest was the comment about `Gens.bools().biasedRepeatingRuns(ratio)` vs `rs.biasedRepeatingRuns(ratio)`; to me this feels like a very different conversation than `Supplier` vs `Gen`.

> This is totally fine when authoring your own classes, but it's less well received by authors of their own classes that they're still maintaining, in which consistency of style will be harmed. 

The code is owned by Apache and we all are expected to work together on it; neither of us *own* any file.  

I still hold by my pushback above, I don't see how I change the style of the code on trunk.  We can not use `RandomSource` directly at each callback as part of the requirement for this JIRA is to add some set of state to improve the randomness, that state is currently stored in a `Gen` and in your other PR you want to introduce `Supplier`.  I would argue that my style is closer to the current style than ours as the *owner* of the `RandomSource` is the top level class.

> I don't think this is the first time we've discussed our conflicting approaches to test authorship, so this shouldn't be a huge surprise? 

You have told me before you don't like `Gen` but have not given actual reasons.  The signature is also present (just different named) in Simulator, so it's hard to see what your argument actually is.

We talked about this in eviction, where I did not force you to match my style, I pointed out that you were duplicating code and had bugs, so you could fix your code (and I showed you how) or reuse.  You had expressed there again that you don't like `Gen` and would prefer to hand roll your own logic, which I was fine with.

> If your style and its components begin to weave their way amongst my style of authorship, then suddenly I have to have opinions about how your API is structured, it's inescapable.

And the same seems to be true for me it seems.  I had expressed my feedback in your patch and am +0 for you to introduce `Supplier`; I don't want to block you because of my preferences, but you actively choose to block me due to yours.  I don't agree with your choice, but I will not block you on it as it is still "correct".  I will ask you to fix or remove code that isn't (which I pointed out in your PR), but since `Supplier` works, I won't block it.  

So yes, we are different here.  I do not try to force you into my style, but you wish to force me, then argue that we should live in peace doing our own things...